### DISCUSSION TITLE
[Question][Hw3] build_call segmantation fault

### DISCUSSION THREAD

**Discussion by cskwak99** [prosyslab-classroom/cs348-information-security#531] [Q&A]
Hello,
I got segmantatio fault when I try ./sanitizer example1.ll

While trying to insert __dbz_sanitizer__, what I did was following
1. declare __dbz_sanitizer using declare_function -> let return value name f
2. make builder using builder_before
3. Try build_call f args "" builder => Segmentation fault occurs

I checked it by print function that build_call is the actual reason for the Segmentation fault.
I tried multiple variants of the above, but couldn't solve it. Can you give me any hints?

+) I also tried to run it in sudo

**Comment by KihongHeo** [comment]
When you ask such a low-level question, first please commit your code. Then it would be much nicer for TAs and instructors to give proper feedback.

**Comment by cskwak99** [comment]
> When you ask such a low-level question, first please commit your code. Then it would be much nicer for TAs and instructors to give proper feedback.

Sorry, now I pushed my commit.

**Comment by RiceBiscuits** [comment]
Please check function type which is used for an argument of declare_function. 
Hint1: A function type contains return type and argument types.
Hint2: Llvm support several API related to types, and also, it contains function type API

**Comment by KihongHeo** [comment]
TA @RiceBiscuits gives a very nice comment. Let me add more general comments for other students also.

You guys may wonder why we have segmentation faults with a safe language, OCaml. LLVM infrastructure is written in C++ and we use an OCaml binding, which is a wrapper of the C++ code. So if you OCaml code incorrectly manipulate the underlying C++ code in a wrong way, you will get segmentation faults. In our homework, common cases may be breaking the correctness of the target C program such as type mismatch, wrong number of arguments, etc, that may cause incorrect object type casting, array index out of bound in C++.

**Comment by cskwak99** [comment]
@RiceBiscuits Thank you very much.
